cart.go

```diff
@@ -13,60 +13,131 @@ import (
 	log.SetOutput(os.Stderr)
 
-	flag.StringVar(&project, "repo", "", "github `username/repo`")
-	flag.StringVar(&branch, "branch", "master", "search builds for branch `name`")
-	flag.IntVar(&buildNum, "build", 0, "get artifact for build number, ignoring branch")
-	flag.StringVar(&workflowArtifactBuild, "workflow-artifact-build", "", "if using a workflow, look for artifacts under build: `name`")
-	flag.IntVar(&workflowDepth, "workflow-depth", defaultWorkflowDepth, "how many workflow builds to check")
 	flag.StringVar(&circleToken, "token", "", "CircleCI auth token")
 	flag.StringVar(&outputPath, "o", "", "output file `path`")
 	flag.BoolVar(&verbose, "v", false, "verbose output")
 	flag.BoolVar(&dryRun, "n", false, "skip artifact download")
 
+	flag.StringVar(&project, "repo", "", "github `username/repo`")
+	flag.IntVar(&buildNum, "build", 0, "get artifact for build number, ignoring branch")
+	flag.StringVar(&filter.branch, "branch", "master", "search builds for branch `name`")
+
+	// Workflows:
+	// If there are multiple workflows, then the latest "build" is perhaps unrelated to building,
+	// not even a later step in a workflow where an earlier step did build.  Eg, we have
+	// stuff to automate dependencies checking, scheduled from cron.
+	// So to retrieve an artifact, we want to only consider specific workflow names.
+	// HOWEVER: those are config items in `.circleci/config.yml` and we should avoid hardcoding
+	// such arbitrary choices across more than one repo, so our default for now is empty,
+	// thus not filtered.
+	//
+	// Within a workflow, the build might not be the last step in the flow; it usually won't be.
+	// Later steps might be "deploy", "stash image somewhere", etc.
+	// So we need to step back from the last step within a workflow until we find the specific
+	// step we're told.
+	//
+	// Eg, for one project, at this time, we use "commit_workflow" as the workflow to search for
+	// and "build" as the job within that workflow.
+
+	flag.StringVar(&filter.workflow, "workflow", "", "only consider builds which are part of this workflow")
+	flag.StringVar(&filter.jobname, "jobname", "", "look within workflow for artifacts from this job's name")
+	flag.IntVar(&retrieveBuildsCount, "retrieve-count", defaultRetrieveCount, "how many builds to retrieve")
+	// This description is too long; how to make it shorter?
+	flag.BoolVar(&filter.anyFlowID, "ignore-later-workflows", false, "get the last successful workflow/job build, even if of a previous flow-id than the latest we see for that workflow")
+
+	// when the workflow-jobname functionality was first added, I (pdp) named it badly; for compatibility,
+	// continue taking the confusingly named option, but map it to the fixed variable.  Similarly for
+	// how the presence of multiple workflows means "workflow-depth" was now a misnomer, and "retrieve-count"
+	// is more accurate.
+	flag.StringVar(&filter.jobname, "workflow-artifact-build", "", "(alias for -workflow-jobname)")
```

Comment:
Does any code use this legacy arg?

Reply:
Yes, the Makefile of the main private-repo app which earns $$$ invokes `cart` with this flag now, so if we lose the flag immediately then branch builds will fail and create unnecessary pain.